Check if the user has truncate privilege on the table

mtuncer opened the pull request:
We rely on postgres to do access control checks for all functions. However, since truncate for cstore tables is handled in process utility,  not forwarded to postgres, and we did not check if a user has truncate right on the table, we let unprivileged user to perform truncate.

This fix introduces acl check prior to truncating files, and reports error if user has failed the check.

Fixes #148

pykello commented on cstore_fdw.c:
```diff
@@ -300,6 +301,9 @@ CStoreProcessUtility(Node * parseTree, const char *queryString,
 		List *cstoreTablesList = FindCStoreTables(allTablesList);
 		List *otherTablesList = list_difference(allTablesList, cstoreTablesList);
+
+		CheckTruncatePermissions(cstoreTablesList);
```
If permissions of a table change by a concurrent session between call to CheckTruncatePermissions() and TruncateCStoreTables(), do we see it here?

(1) PostgreSQL's ExecuteTruncate() seems to do permission check & truncation while holding an AccessExclusiveLock, so I guess permissions cannot change after permission check and before end of TRUNCATE.
(2) Not caused by this change, but PostgreSQL grabs a lock on *all* relations before starting truncating the first table. In cstore_fdw we lock, truncate, release tables one by one.

So our behaviour here is not similar to PostgreSQL here.  I don't think this is a big issue, just trying to understand.

@anarazel what do you think? Can not having same semantics as PostgreSQL here cause much harm? (I'm sure these 2 aren't the only issues)

anarazel replied:
> (1) PostgreSQL's ExecuteTruncate() seems to do permission check & truncation while holding an AccessExclusiveLock, so I guess permissions cannot change after permission check and before end of TRUNCATE.

I'm not particularly concerned about that race. But we really *should* take an AccessExclusiveLock during the check (or just acquire one before), because otherwise there's lock-upgrade issues.

>  (2) Not caused by this change, but PostgreSQL grabs a lock on all relations before starting truncating the first table. In cstore_fdw we lock, truncate, release tables one by one.

I don't think we "release" locks (that's what you mean, right) at all?

pykello replied:
In TruncateCStoreTable() in cstore_fdw.c, doesn't heap_open() acquire a lock and heap_close() release that lock?

anarazel replied:
> In TruncateCStoreTable() in cstore_fdw.c, doesn't heap_open() acquire a lock and heap_close() release that lock?

Indeed. And that's completely wrong.  I was only looking at CheckTruncatePermissions() which acquires a lock (of the too low level), and doesn't release it.

mtuncer commented:
Changed locking semantics
- renamed ACL checking function so that it checks access control, locks with AccessExclusiveLock, and returns relation list.
- Truncate functions now accepts list of previously locked relations, truncates, and unlocks them.

pykello commented on cstore_fdw.c:
```diff
@@ -786,20 +789,63 @@ FindCStoreTables(List *tableList)
+ * on tables.
+ */
+static List *
+OpenTruncatedRelations(List *cstoreTableList)
```
naming suggestion:‌ OpenRelationsForTruncate()?

mtuncer replied:
sounds better

pykello commented on cstore_fdw.c:
```diff
@@ -358,7 +361,7 @@ CStoreProcessUtility(Node * parseTree, const char *queryString,
 		}
 
-		TruncateCStoreTables(cstoreTablesList);
+		TruncateCStoreTables(cstoreRelationList);
```
`heap_close()` the relations after truncation? (See `ExecuteTruncate()` in tablecmds.c)

mtuncer replied:
```TruncateCStoreTables()``` does heap_close. Do you mean closing relations outside ```TruncateCStoreTables()``` ?

pykello replied:
Sorry, I‌ had missed that. Maybe we should close outside `TruncateCStoreTables()` as you suggested so future readers of code don't miss that.

pykello commented:
Looks good to me. We can merge.
